From c1ab3304d3ccf68730b1040400b9497734e860fe Mon Sep 17 00:00:00 2001
From: David Holsgrove <david.holsgrove@xilinx.com>
Date: Tue, 20 Nov 2012 13:00:08 +1000
Subject: [PATCH 68/79] microblaze: Cannot emit swap insns if -mno-xl-reorder passed on cmd line

The content of this patch will be merged into the initial swap instruction
patch for upstream submission to gcc mainline.

Signed-off-by: David Holsgrove <david.holsgrove@xilinx.com>
---
 gcc/config/microblaze/microblaze.c   |   22 ++++++++++++++++++----
 gcc/config/microblaze/microblaze.opt |    6 +++++-
 2 files changed, 23 insertions(+), 5 deletions(-)

diff --git a/gcc/config/microblaze/microblaze.c b/gcc/config/microblaze/microblaze.c
index 62d4750..05750f5 100755
--- a/gcc/config/microblaze/microblaze.c
+++ b/gcc/config/microblaze/microblaze.c
@@ -160,6 +160,9 @@ int microblaze_has_clz = 0;
 /* Set to one if the targeted core has the swapb and swaph insn.  */
 int microblaze_has_swap = 0;
 
+/* Set to one if -mno-xl-reorder has been passed.  */
+int microblaze_no_reorder = 0;
+
 /* Which CPU pipeline do we use. We haven't really standardized on a CPU 
    version having only a particular type of pipeline. There can still be 
    options on the CPU to scale pipeline features up or down. :( 
@@ -1732,6 +1735,9 @@ microblaze_handle_option (size_t code,
     case OPT_mxl_stack_check:
       warning (0, "-mxl_stack_check is deprecated; use -fstack-check");
       break;
+    case OPT_mno_xl_reorder:
+      microblaze_no_reorder = 1;
+      break;
     }
   return true;
 }
@@ -1839,15 +1845,23 @@ microblaze_option_override (void)
         /* MicroBlaze v8.30a has an undocumented dependency on pattern compare for swapb / swaph insns. */
         if (TARGET_PATTERN_COMPARE)
         {
-            microblaze_has_swap = 1;
-            target_flags |= MASK_REORDER;
+            /* Cannot add MASK_REORDER or use swap insns if -mno-xl-reorder passed */
+            if (!microblaze_no_reorder)
+            {
+                target_flags |= MASK_REORDER;
+                microblaze_has_swap = 1;
+            }
         }
     }
   else
    {
         /* Microblaze versions greater than v8.30a will be able to use swapb / swaph without pattern compare */
-        microblaze_has_swap = 1;
-        target_flags |= MASK_REORDER;
+        /* Cannot add MASK_REORDER or use swap insns if -mno-xl-reorder passed */
+        if (!microblaze_no_reorder)
+        {
+            target_flags |= MASK_REORDER;
+            microblaze_has_swap = 1;
+        }
    }
 
   if (TARGET_MULTIPLY_HIGH && TARGET_SOFT_MUL)
diff --git a/gcc/config/microblaze/microblaze.opt b/gcc/config/microblaze/microblaze.opt
index 426ee7a..6c5a020 100755
--- a/gcc/config/microblaze/microblaze.opt
+++ b/gcc/config/microblaze/microblaze.opt
@@ -68,9 +68,13 @@ Target Mask(SOFT_MUL)
 Use the soft multiply emulation (default)
 
 mxl-reorder
-Target Mask(REORDER)
+Target RejectNegative Mask(REORDER)
 Use reorder instructions (default)
 
+mno-xl-reorder
+Target RejectNegative
+Do not use reorder instructions
+
 mxl-soft-div
 Target Mask(SOFT_DIV)
 Use the software emulation for divides (default)
-- 
1.7.1

